src/ripple/app/paths/AMMOffer.h

```diff
@@ -0,0 +1,143 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2023 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_AMMOFFER_H_INCLUDED
+#define RIPPLE_APP_AMMOFFER_H_INCLUDED
+
+#include <ripple/ledger/ApplyView.h>
+#include <ripple/protocol/Quality.h>
+
+namespace ripple {
+
+template <typename TIn, typename TOut>
+class AMMLiquidity;
+class QualityFunction;
+
+/** Represents synthetic AMM offer in BookStep. AMMOffer mirrors TOffer
+ * methods for use in generic BookStep methods. AMMOffer amounts
+ * are changed indirectly in BookStep limiting steps.
+ */
+template <typename TIn, typename TOut>
+class AMMOffer
+{
+private:
+    AMMLiquidity<TIn, TOut> const& ammLiquidity_;
+    // Initial offer amounts. It is fibonacci seq generated for multi-path.
+    // If the offer size is set based on the competing CLOB offer then
+    // the AMM offer size is such that if the offer is consumed then
+    // the updated AMM pool SP quality is going to be equal to competing
+    // CLOB offer quality. If there is no competing CLOB offer then
+    // the initial size is set to in=cMax[Native,Value],balances.out.
+    // While this is not a "real" offer it simulates the case of
+    // the swap out of the entire side of the pool, in which case
+    // the swap in amount is infinite.
+    TAmounts<TIn, TOut> const amounts_;
+    // If seated then current pool balances. Used in one-path limiting steps
+    // to swap in/out.
+    std::optional<TAmounts<TIn, TOut>> const balances_;
+    // The Spot Price quality if balances != amounts
+    // else the amounts quality
+    Quality const quality_;
+    // AMM offer can be consumed once at a given iteration
+    bool consumed_;
+
+public:
+    AMMOffer(
+        AMMLiquidity<TIn, TOut> const& ammLiquidity,
+        TAmounts<TIn, TOut> const& amounts,
+        std::optional<TAmounts<TIn, TOut>> const& balances,
+        Quality const& quality);
+
+    Quality const
+    quality() const noexcept
```

Comment:
If you're intending to make a copy (which looks like the case), then the `const` is pointless - the caller will make a copy when assigning. If you're not trying to make a copy, then it might make sense to return `Quality const&`.

(Normally, I'd say return the `const&` regardless, but it looks like the size of a `Quality` is the same as the size of a pointer, so there's no direct benefit.)